Duet AI Prompts - Documentation Lookup Without Links (#30873)
apache/beam: learning/prompts/documentation-lookup-nolinks/30_advanced_additional_outputs.md

  PCollection<String> input = ...;
// Output tag for the main output PCollection of strings.
final TupleTag<String> mainOutputTag = new TupleTag<String>(){};
// Output tag for the additional output PCollection of strings.
final TupleTag<String> additionalOutputTagString = new TupleTag<Integer>(){};
// Output tag for the additional output PCollection of integers.
final TupleTag<Integer> additionalOutputTagIntegers = new TupleTag<Integer>(){};

PCollectionTuple results = input.apply(ParDo
          .of(new DoFn<String, String>() {
            // DoFn continues here.
            ...
          })
          // Specify the tag for the main output.
          .withOutputTags(mainOutputTag,
          // Specify the tags for the two additional outputs as a TupleTagList.
                          TupleTagList.of(additionalOutputTagString)
                                      .and(additionalOutputTagIntegers)));
```

The `processElement` method can emit elements to the main output or any additional output by invoking the output method on the `MultiOutputReceiver` object. The output method takes the tag of the output and the element to be emitted as arguments.

```java

> public void processElement(@Element String word, MultiOutputReceiver out) {
>        if (condition for main output) {
>          // Emit element to main output
>          out.get(mainOutputTag).output(word);
>        } else {
>          // Emit element to additional string output
>          out.get(additionalOutputTagString).output(word);
>        }
>        if (condition for additional integer output) {
>          // Emit element to additional integer output
>          out.get(additionalOutputTagIntegers).output(word.length());
>        }
>      }

Reviewer: ```suggestion
public void processElement(@Element String word, MultiOutputReceiver out) {
    if (condition
        for main output) {
        // Emit element to main output
        out.get(mainOutputTag).output(word);
    } else {
        // Emit element to additional string output
        out.get(additionalOutputTagString).output(word);
    }
    if (condition
        for additional integer output) {
        // Emit element to additional integer output
        out.get(additionalOutputTagIntegers).output(word.length());
    }
}
```

```

In the Apache Beam Python SDK, you can implement additional outputs by invoking the `with_outputs()` method on the `ParDo` transform and specifying the expected tags for the multiple outputs.

The following Python code demonstrates how to implement additional outputs for a `ParDo` transform that outputs two `PCollection` objects of strings and integers in addition to the main output `PCollection` of strings:

```python

> class SplitLinesToWordsFn(beam.DoFn):
> 
>   # These tags will be used to tag the outputs of this DoFn.
>   OUTPUT_TAG_SHORT_WORDS = 'tag_short_words'
>   OUTPUT_TAG_CHARACTER_COUNT = 'tag_character_count'
> 
>   def process(self, element):
>     # yield a count (integer) to the OUTPUT_TAG_CHARACTER_COUNT tagged collection.
>     yield pvalue.TaggedOutput(self.OUTPUT_TAG_CHARACTER_COUNT, len(element))
> 
>     words = re.findall(r'[A-Za-z\']+', element)
>     for word in words:
>       if len(word) <= 3:
>         # yield word as an output to the OUTPUT_TAG_SHORT_WORDS tagged collection.
>         yield pvalue.TaggedOutput(self.OUTPUT_TAG_SHORT_WORDS, word)
>       else:
>         # yield word to add it to the main collection.
>         yield word

Reviewer: ```suggestion
class SplitLinesToWordsFn(beam.DoFn):
    # These tags will be used to tag the outputs of this DoFn.
    OUTPUT_TAG_SHORT_WORDS = "tag_short_words"
    OUTPUT_TAG_CHARACTER_COUNT = "tag_character_count"

    def process(self, element):
        # yield a count (integer) to the OUTPUT_TAG_CHARACTER_COUNT tagged collection.
        yield pvalue.TaggedOutput(self.OUTPUT_TAG_CHARACTER_COUNT, len(element))

        words = re.findall(r"[A-Za-z\']+", element)
        for word in words:
            if len(word) <= 3:
                # yield word as an output to the OUTPUT_TAG_SHORT_WORDS tagged collection.
                yield pvalue.TaggedOutput(self.OUTPUT_TAG_SHORT_WORDS, word)
            else:
                # yield word to add it to the main collection.
                yield word
```

```

The method returns a `DoOutputsTuple` object, with the specified tags serving as attributes that provide `ParDo` with access to the corresponding output `PCollection` objects.

```python

> with beam.Pipeline(options=pipeline_options) as p:
> 
>     lines = p | ReadFromText(known_args.input)
> 
>     # with_outputs allows accessing the explicitly tagged outputs of a DoFn.
>     split_lines_result = (
>         lines
>         | beam.ParDo(SplitLinesToWordsFn()).with_outputs(
>             SplitLinesToWordsFn.OUTPUT_TAG_SHORT_WORDS,
>             SplitLinesToWordsFn.OUTPUT_TAG_CHARACTER_COUNT,
>             main='words'))
> 
>     # split_lines_result is an object of type DoOutputsTuple
>     words, _, _ = split_lines_result
>     short_words = split_lines_result[SplitLinesToWordsFn.OUTPUT_TAG_SHORT_WORDS]
>     character_count = split_lines_result.tag_character_count

Reviewer: ```suggestion
with beam.Pipeline(options=pipeline_options) as p:
    lines = p | ReadFromText(known_args.input)

    # with_outputs allows accessing the explicitly tagged outputs of a DoFn.
    split_lines_result = lines | beam.ParDo(SplitLinesToWordsFn()).with_outputs(
        SplitLinesToWordsFn.OUTPUT_TAG_SHORT_WORDS,
        SplitLinesToWordsFn.OUTPUT_TAG_CHARACTER_COUNT,
        main="words",
    )

    # split_lines_result is an object of type DoOutputsTuple
    words, _, _ = split_lines_result
    short_words = split_lines_result[SplitLinesToWordsFn.OUTPUT_TAG_SHORT_WORDS]
    character_count = split_lines_result.tag_character_count
```
